README.md
# Beholder
```
Beholder V1.10.001 - ELK/BRO/Libtrace
Created By: Jason Azzarella and Chris Pavan
Problems or Feature Requests?
E-mail Us: jmazzare@bechtel.com
```
Beholder is a shell script which installs and configures essentials to peer into your network activity. Monitor your network traffic with Bro IDS, build dashboards with Kibana to get a visual representation of your activity and obtain packet captures of the identified events with Tracesplit.

## Under the Hood

- ELK Stack (https://www.elastic.co)
  - Elasticsearch
    - Curator
  - Logstash
  - Kibana
- Bro IDS (https://www.bro.org/)
- Libtrace (http://research.wand.net.nz/software/libtrace.php)

## Software Requirements

- Ubuntu 16.04 x64

## Hardware Minimum Requirements

- 64 Bit Processor
- 3 GB RAM
- 40 GB HDD

## Installation

- Run the beholder script. Use sudo ./beholder.
- Select the interface and IP address for monitoring and management.
- At completion, the system will countdown and reboot.
- Keep watch for the special message!

## Details

- Linux User beholder
  - Pass beholder
- Basic Auth User beholder
  - Pass beholder
- Kibana Interface - https://{Your Kibana IP}
  1. Use a web browser to access your Kibana instance.
  2. On your first load, you will be required to insert the elasticsearch index.
  3. Input bro* and select the @timestamp field from the timestamp dropdown.
  4. You will be presented with the index fields that have been identified. Use the yellow refresh button in order to update the listed fields. You may need to hit this a few times on first setup.
  5. You should get over 200 fields after all have been populated at least once.

- Tracesplit Example
  - Capture data from interface eth0 - sudo /opt/libtrace/bin/tracesplit -z 6 -Z gzip int:eth0 erf:/pcaps/capture.gz
	
## What's in a Version Name?

Version numbers get confusing so I'm adding an explanation as to what matters for versions of this script.

EX 1.00.001

- The first set of digits (1) represents major tool additions and new functionality.
- The second set of digits (00) represents upgraded versions of tools such as Bro, Elasticsearch, Logstash.
- The third set of digits (001) represents tweaks to the scripts, changes to the templates or all around fixes to the code.


ChangeLog.txt
Change Log

------------------
Beholder V1.10.001
------------------

Changes

* bro_http fields fixed in the logstash config file.

------------------
Beholder V1.10.000
------------------
Current Versions

* ElasticSearch: 5.4 - Upgraded
* Logstash: 5.4 - Upgraded
* Kibana: 5.4 - Upgraded
* Curator - Added

Changes

* Upgraded to new ELK version.
* Added Curator functionality after the 5.x upgrades.
 * Bro indices older than 30 days are now closed once a day via cron.

------------------
Beholder V1.09.000
------------------
Current Versions

* ElasticSearch: 5.0 - Upgraded
* Logstash: 5.0 - Upgraded
* Kibana: 5.0 - Upgraded
* Bro: 2.5 - Upgraded

Changes

* Upgraded to be used on Ubuntu 16!
* Changed autostart functionality from upstart to systemd.
* ELK and Bro upgraded to the newest version.
* File is now executable. Install should be done as sudo ./beholder.
* Now Proxy capable!
* All input requirements are now upfront.
  * You can now select a different IP/interface between Bro and the ELK bound IP.
* Curator is no longer auto-configured for closing out elk indexes.
* Removed some of the app clutter from app-get install.

------------------
Beholder V1.08.000
------------------
Current Versions

* ElasticSearch: 2.3.2 - Upgraded
* Logstash: 2.3.2 - Upgraded
* Kibana: 4.5.0 - Upgraded
* Bro: 2.4.1

Feature Changes

* Added 30 other log sources from Bro which feeds into the ELK stack. This totals up to about 37.
* A lot of fields have been normalized. Any "bytes" field has been changed to "bytes_(x)" to make it easier to poke through.

Fixes

* Changed the filter from Logstash which should remove any previously failed parses.

------------------
Beholder V1.07.000
------------------
Current Versions

* ElasticSearch: 2.2.1
* Logstash: 2.2.2
* Kibana: 4.4.2
* Bro: 2.4.1

Feature Changes

* Script now contains Apache and Kibana is served up via https. No longer required to hit 5601.
* Apache sets itself up with a self-signed certificate.
* Basic Auth now implemented to access Kibana.
* UFW turned on with default allow and blocking kibana from external.
* Special Message added to the script.

Fixes

* Beholder no longer added to sudo.
* Bit of changes to the script so that there is less interaction required from the user.
* User and Pass beholder/beholder for both basic auth and Linux user.

------------------
Beholder V1.06.000
------------------
Current Versions

* ElasticSearch: 2.2.1 - Updated
* Logstash: 2.2.2 - Updated
* Kibana: 4.4.2 - Updated
* Bro: 2.4.1

Feature Changes

NA

Fixes

* Added beholder HOME directory.
* Tossed the ELK stack into a dropbox location as there tends to be frequent updates and this will allow the script to always be relevant.

------------------
Beholder V1.05.000
------------------
Current Versions

* ElasticSearch: 2.1.1 - Updated
* Logstash: 2.1.1 - Updated
* Kibana: 4.3.1 - Updated
* Bro: 2.4.1

Feature Changes

NA

------------------
Beholder V1.03.001
------------------
Current Versions

* ElasticSearch: 1.7.2
* Logstash: 1.5.4
* Kibana: 4.1.2
* Bro: 2.4.1

Feature Changes

* Ubuntu 15 is now supported.
	* Script will check your Ubuntu version and install necessary extra packages to get init running.
* INIT
	* "Stoping" is now "stopping"... Yay!

------------------
Beholder V1.03.000
------------------
Current Versions

* ElasticSearch: 1.7.2 - Updated
* Logstash: 1.5.4
* Kibana: 4.1.2 - Updated
* Bro: 2.4.1 - Updated

------------------
Beholder V1.02.000
------------------
Current Versions

* ElasticSearch: 1.7.1 - Updated
* Logstash: 1.5.4 - Updated
* Kibana: 4.1.1
* Bro: 2.4

Feature Changes

* Cron
	* Fixed an issue with curator running from cron.

------------------
Beholder V1.01.000
------------------
Current Versions

* ElasticSearch: 1.7.0 - Updated
* Logstash: 1.5.3 - Updated
* Kibana: 4.1.1
* Bro: 2.4

Feature Changes

* Elasticsearch
	* Curator
		* Added "--prefix bro" to the cronjob so that custom indices will not be caught.

------------------
Beholder V1.00.001
------------------
Current Versions

* ElasticSearch: 1.6.0
* Logstash: 1.5.2
* Kibana: 4.1.1
* Bro: 2.4

Feature Changes

* Elasticsearch
	* Added Integer type fields for the duration and bytes fields from Bro. 
	
* Kibana
	* Because of the change in field types, SUM/Min/Max for Kibana dashboards can now properly function against the bytes and duration field.

Bug Fix

* Bro
	* Unable to compile Git code
		* Changed from getting Bro from git to pulling from their website.


beholder.sh
#!/bin/bash
############################################
#Beholder V1.07.000 - ELK/BRO/Libtrace
#Created By: Jason Azzarella and Chris Pavan
#Problems or Feature Requests?
#E-mail Us: jmazzare@bechtel.com
############################################
clear
dt() {
date -u '+%m/%d/%Y %H:%M:%S'
}
rootcheck() {
	if [ $(id -u) != "0" ]; then
		echo -e "$(dt) Elevating your privileges..."
		sudo "sh" "$0" "$@"
		exit $?
	fi
}
rootcheck
#####################
#Check Ubuntu Version
#####################
echo "[+] Ubuntu Version Check."
apt-get update
apt-get install -y lsb-core
version=$(lsb_release -a | grep Release | awk '{print $2}' | sed 's/\..*//')
versioncheck() {
	if [ $version = "15" ]; then
		echo "You are on Ubuntu:" $version
		echo "Your Ubuntu version is supported. Installing init support."
		apt-get install -y upstart-sysv openjdk-8-jre
		update-initramfs -u
	elif [ $version = "14" ]; then
		sudo add-apt-repository ppa:webupd8team/java -y
		sudo apt-get update
		echo oracle-java8-installer shared/accepted-oracle-license-v1-1 select true | /usr/bin/debconf-set-selections
		sudo apt-get install oracle-java8-installer -y
		echo "You are on Ubuntu:" $version
		echo "Your Ubuntu version is supported."
	else
		echo "Beholder does not support version" $version
		echo "Exiting Beholder."
		exit
	fi
}
versioncheck
clear
#######################
#Creating Beholder User
#######################
echo "[+] Creating beholder user."
useradd beholder -m -d /home/beholder
echo 'beholder:beholder' | chpasswd
#####################
#Build File Structure
#####################
echo "[+] Setting up the file system."
mkdir /logs
mkdir /logs/bro
mkdir /logs/elasticsearch
mkdir /logs/index
mkdir /logs/bro/spool
mkdir /logs/logstash
mkdir /pcaps/
mkdir /home/beholder
chown beholder:beholder /home/beholder
#####################################
#Installing Updates and Dependencies.
#####################################
echo "[+] Starting download and install. This WILL take a while. Be cool!"
wget -qO - https://packages.elasticsearch.org/GPG-KEY-elasticsearch | sudo apt-key add -
cd /etc/apt/sources.list.d/
cat <<EOF > curator.list
deb http://packages.elasticsearch.org/curator/3/debian stable main
EOF
apt-get update
apt-get install -y apache2 apache2-utils unzip bless lsb-core cmake make gcc g++ flex bison libpcap-dev libssl-dev python-dev swig zlib1g-dev git dh-autoreconf python-elasticsearch-curator
#####################
#Installing ELK Stack
#####################
echo "[+] Installing ELK Stack"
cd /opt/
wget https://www.dropbox.com/s/r27jwgo13d0hnga/elasticsearch-2.2.1.tar.gz
tar -zxvf *.tar.gz
rm -rf *.tar.gz
mv elastic* elasticsearch
wget https://www.dropbox.com/s/gz4txuatpcl5sbh/logstash-2.2.2.tar.gz
tar -zxvf *.tar.gz
rm -rf *.tar.gz
mv logstash* logstash
wget https://www.dropbox.com/s/pv9xj6dl2sa8mk8/kibana-4.4.2-linux-x64.tar.gz
tar -zxvf *.tar.gz
rm -rf *.tar.gz
mv kibana-* kibana
###############
#Installing Bro
###############
echo "[+] Installing Bro"
cd /opt/
mkdir /opt/broinstall
mkdir /opt/bro
cd /opt/broinstall
wget https://www.bro.org/downloads/release/bro-2.4.1.tar.gz
tar -zxvf bro*
cd /opt/broinstall/bro-2.4.1
./configure --prefix=/opt/bro
make
make install
rm -rf /opt/broinst*
####################
#Installing Libtrace
####################
echo "[+] Installing Libtrace"
cd /opt/
wget http://research.wand.net.nz/software/libtrace/libtrace-latest.tar.bz2
tar jxf libtrace-latest.tar.bz2
rm -rf libtrace-latest.tar.bz2
mv libtrace-* libtrace
cd /opt/libtrace
./configure
make
make install
###############
#Configurations
###############
echo "[+] Beginning Configurations"
###############################
#Configuration - Bro Node Setup
###############################
clear
echo "Bro is used to monitor traffic on an interface."
echo "We have identified the following interfaces on your system;"
echo "-----------------------------------------------------------"
for line in $(awk 'NR>2 {print $1}' /proc/net/dev | sed 's/://g'); do
        echo "$line"
done
echo "-----------------------------------------------------------"
read -p "Which interface above would you like to monitor: " broinput
sed -i "s/interface=.*/interface=$broinput/" /opt/bro/etc/node.cfg
##############################
#Configuration - Bro Logs Path
##############################
sed -i 's/LogDir\s=.*/LogDir = \/logs\/bro/' /opt/bro/etc/broctl.cfg
sed -i 's/SpoolDir\s=.*/SpoolDir = \/logs\/bro\/spool/' /opt/bro/etc/broctl.cfg
############################
#Configuration - Bro Install
############################
/opt/bro/bin/broctl install
/opt/bro/bin/broctl deploy
##############################
#Configuration - Elasticsearch
##############################
cd /opt/elasticsearch/config
cat <<EOF > elasticsearch.yml
cluster.name: beholder
node.name: beholder
path.data: /logs/index
path.logs: /logs/elasticsearch
EOF
################################
#Configuration - Logstash Inputs
################################
mkdir /opt/logstash/config
cd /opt/logstash/config
#####################################
#Configuration - Logstash ES Template
#####################################
cat <<EOF > /opt/logstash/config/bro.json
{
    "template": "bro*",
    "settings": {
        "index.refresh_interval": "5s"
    },
    "mappings": {
        "_default_": {
            "_all": {
                "enabled": true
            },
            "dynamic_templates": [
                {
                    "string_fields": {
                        "match": "*",
                        "match_mapping_type": "string",
                        "mapping": {
                            "type": "string",
                            "index": "analyzed",
                            "omit_norms": true,
                            "fields": {
                                "raw": {
                                    "type": "string",
                                    "index": "not_analyzed",
                                    "ignore_above": 1024
                                }
                            }
                        }
                    }
                }
            ],
            "properties": {
                "@version": {
                    "type": "string",
                    "index": "not_analyzed"
                },
                "dest_ip_bytes": {
                    "type": "integer",
                    "ignore_malformed": true,
                    "index": "analyzed"
                },
                "dest_bytes": {
                    "type": "integer",
                    "ignore_malformed": true,
                    "index": "analyzed"
                },
                "dest_pkts": {
                    "type": "integer",
                    "ignore_malformed": true,
                    "index": "analyzed"
                },
                "duration": {
                    "type": "float",
                    "ignore_malformed": true,
                    "index": "analyzed"
                },
                "missed_bytes": {
                    "type": "integer",
                    "ignore_malformed": true,
                    "index": "analyzed"
                },
                "missing_bytes": {
                    "type": "integer",
                    "ignore_malformed": true,
                    "index": "analyzed"
                },
                "seen_bytes": {
                    "type": "integer",
                    "ignore_malformed": true,
                    "index": "analyzed"
                },
                "src_bytes": {
                    "type": "integer",
                    "ignore_malformed": true,
                    "index": "analyzed"
                },
                "src_ip_bytes": {
                    "type": "integer",
                    "ignore_malformed": true,
                    "index": "analyzed"
                },
                "src_pkts": {
                    "type": "integer",
                    "ignore_malformed": true,
                    "index": "analyzed"
                },
                "total_bytes": {
                    "type": "integer",
                    "ignore_malformed": true,
                    "index": "analyzed"
                }
            }
        }
    }
}
EOF
####################################
#Configuration - Logstash Bro Parser
####################################
cat <<EOF > /opt/logstash/config/bro.conf
input {
        file {
                path => "/logs/bro/spool/bro/files.log"
                type => "bro_files"
                sincedb_path => "/logs/logstash/brofiles"
        }
        file {
                path => "/logs/bro/spool/bro/dhcp.log"
                type => "bro_dhcp"
                sincedb_path => "/logs/logstash/brodhcp"
        }
        file {
                path => "/logs/bro/spool/bro/http.log"
                type => "bro_http"
                sincedb_path => "/logs/logstash/brohttp"
        }
        file {
                path => "/logs/bro/spool/bro/ssl.log"
                type => "bro_ssl"
                sincedb_path => "/logs/logstash/brossl"
        }
        file {
                path => "/logs/bro/spool/bro/dns.log"
                type => "bro_dns"
                sincedb_path => "/logs/logstash/brodns"
        }
        file {
                path => "/logs/bro/spool/bro/conn.log"
                type => "bro_conn"
                sincedb_path => "/logs/logstash/broconn"
        }
        file {
                path => "/logs/bro/spool/bro/smtp.log"
                type => "bro_smtp"
                sincedb_path => "/logs/logstash/brosmtp"
        }
}
filter {
		if ([message] =~ /^#/) {
        	drop{}
    	}
        else if [type] == "bro_files" {
                grok {
                        match => [ "message", "(?<time>(.*?))\t(?<bro_id>(.*?))\t(?<source_ip>(.*?))\t(?<dest_ip>(.*?))\t(?<conn_uids>(.*?))\t(?<source>(.*?))\t(?<depth>(.*?))\t(?<analyzers>(.*?))\t(?<mime_type>(.*?))\t(?<filename>(.*?))\t(?<duration>(.*?))\t(?<local_orig>(.*?))\t(?<is_orig>(.*?))\t(?<seen_bytes>(.*?))\t(?<total_bytes>(.*?))\t(?<missing_bytes>(.*?))\t(?<overflow_bytes>(.*?))\t(?<timedout>(.*?))\t(?<parent_fuid>(.*?))\t(?<md5>(.*?))\t(?<sha1>(.*?))\t(?<sha256>(.*?))\t(?<extracted>(.*?))" ]
                        }
        }
        else if [type] == "bro_dhcp" {
                grok {
                        match => [ "message", "(?<time>(.*?))\t(?<bro_id>(.*?))\t(?<source_ip>(.*?))\t(?<source_port>(.*?))\t(?<dest_ip>(.*?))\t(?<dest_port>(.*?))\t(?<mac>(.*?))\t(?<assigned_ip>(.*?))\t(?<lease_time>(.*?))\t(?<trans_id>(.*?))" ]
                        }
        }
        else if [type] == "bro_http" {
                grok {
                        match => [ "message", "(?<time>(.*?))\t(?<bro_id>(.*?))\t(?<source_ip>(.*?))\t(?<source_port>(.*?))\t(?<dest_ip>(.*?))\t(?<dest_port>(.*?))\t(?<depth>(.*?))\t(?<method>(.*?))\t(?<domain>(.*?))\t(?<uri>(.*?))\t(?<referrer>(.*?))\t(?<user_agent>(.*?))\t(?<request_length>(.*?))\t(?<response_length>(.*?))\t(?<status_code>(.*?))\t(?<status_msg>(.*?))\t(?<info_code>(.*?))\t(?<info_msg>(.*?))\t(?<filename>(.*?))\t(?<tags>(.*?))\t(?<username>(.*?))\t(?<password>(.*?))\t(?<proxied>(.*?))\t(?<orig_fuids>(.*?))\t(?<orig_mime_types>(.*?))\t(?<resp_fuids>(.*?))\t(?<resp_mime_types>(.*?))" ]
                        }
        }
        else if [type] == "bro_ssl" {
                grok {
                        match => [ "message", "(?<time>(.*?))\t(?<bro_id>(.*?))\t(?<source_ip>(.*?))\t(?<source_port>(.*?))\t(?<dest_ip>(.*?))\t(?<dest_port>(.*?))\t(?<version>(.*?))\t(?<cypher>(.*?))\t(?<curve>(.*?))\t(?<server_name>(.*?))\t(?<resumed>(.*?))\t(?<last_alert>(.*?))\t(?<next_protocol>(.*?))\t(?<established>(.*?))\t(?<cert_chain_fuids>(.*?))\t(?<client_cert_chain_fuids>(.*?))\t(?<subject>(.*?))\t(?<issuer>(.*?))\t(?<client_subject>(.*?))\t(?<client_issuer>(.*?))\t(?<validation_status>(.*?))" ]
                        }
        }
        else if [type] == "bro_dns" {
                grok {
                        match => [ "message", "(?<time>(.*?))\t(?<bro_id>(.*?))\t(?<source_ip>(.*?))\t(?<source_port>(.*?))\t(?<dest_ip>(.*?))\t(?<dest_port>(.*?))\t(?<proto>(.*?))\t(?<trans_id>(.*?))\t(?<query>(.*?))\t(?<qclass>(.*?))\t(?<qclass_name>(.*?))\t(?<qtype>(.*?))\t(?<qtype_name>(.*?))\t(?<rcode>(.*?))\t(?<rcode_name>(.*?))\t(?<AA>(.*?))\t(?<TC>(.*?))\t(?<RD>(.*?))\t(?<RA>(.*?))\t(?<Z>(.*?))\t(?<answers>(.*?))\t(?<ttls>(.*?))\t(?<rejected>(.*?))" ]
                        }
        }
        else if [type] == "bro_conn" {
                grok {
                        match => [ "message", "(?<time>(.*?))\t(?<bro_id>(.*?))\t(?<source_ip>(.*?))\t(?<source_port>(.*?))\t(?<dest_ip>(.*?))\t(?<dest_port>(.*?))\t(?<proto>(.*?))\t(?<service>(.*?))\t(?<duration>(.*?))\t(?<src_bytes>(.*?))\t(?<dest_bytes>(.*?))\t(?<state>(.*?))\t(?<local_orig>(.*?))\t(?<local_resp>(.*?))\t(?<missed_bytes>(.*?))\t(?<history>(.*?))\t(?<src_pkts>(.*?))\t(?<src_ip_bytes>(.*?))\t(?<dest_pkts>(.*?))\t(?<dest_ip_bytes>(.*?))\t(?<tunnel_parents>(.*?))" ]
                        }
        }
        else if [type] == "bro_smtp" {
                grok {
                        match => [ "message", "(?<time>(.*?))\t(?<bro_id>(.*?))\t(?<id.orig_h>(.*?))\t(?<id.resp_h>(.*?))\t(?<id.resp_p>(.*?))\t(?<trans_depth>(.*?))\t(?<helo>(.*?))\t(?<mailfrom>(.*?))\t(?<rcptto>(.*?))\t(?<date>(.*?))\t(?<from>(.*?))\t(?<to>(.*?))\t(?<reply_to>(.*?))\t(?<msg_id>(.*?))\t(?<in_reply_to>(.*?))\t(?<subject>(.*?))\t(?<x_originating_ip>(.*?))\t(?<first_received>(.*?))\t(?<second_received>(.*?))\t(?<last_reply>(.*?))\t(?<path>(.*?))\t(?<user_agent>(.*?))" ]
                		}
        }
}
output {
        elasticsearch {
                hosts => ["localhost:9200"]
                index => "bro-%{+YYYY.MM.dd}"
                template => "/opt/logstash/config/bro.json"
                template_name => "bro*"
        }
        stdout {
                codec => rubydebug
        }
}
EOF
#########################
#Setup Initialize Scripts
#########################
echo "[+] Setting up Init Scripts"
##############
#Logstash Init
##############
cd /etc/init.d
cat <<'EOF' > logstash
. /lib/lsb/init-functions
name="logstash"
logstash_bin="-- /opt/logstash/bin/logstash"
logstash_conf="/opt/logstash/config/bro.conf"
logstash_log="/logs/logstash/$name.log"
pid_file="/var/run/$name.pid"
NICE_LEVEL="-n 19"
HOME=/home/beholder
start () {
    command="/usr/bin/nice ${NICE_LEVEL} ${logstash_bin} agent -f $logstash_conf --log ${logstash_log} -- web"

    log_daemon_msg "Starting $name"
    if start-stop-daemon --start --chuid "beholder" --quiet --oknodo --pidfile "$pid_file" -b -m --exec $command; then
        log_end_msg 0
    else
        log_end_msg 1
    fi
}
stop () {
    echo "Stopping $name"
    start-stop-daemon --stop --quiet --oknodo --pidfile "$pid_file"
    echo "$name stopped"
}

status () {
    status_of_proc -p $pid_file "" "$name"
}
case $1 in
    start)
        if status; then exit 0; fi
        start
        ;;
    stop)
        stop
        ;;
    reload)
        stop
        start
        ;;
    restart)
        stop
        start
        ;;
    status)
        status && exit 0 || exit $?
        ;;
    *)
        echo "Usage: $0 {start|stop|restart|reload|status}"
        exit 1
        ;;
esac
exit 0
EOF
chmod +x logstash
update-rc.d logstash defaults
###################
#Elasticsearch Init
###################
cd /etc/init.d
cat <<'EOF' > elasticsearch
. /lib/lsb/init-functions
name="elasticsearch"
elastic="-- /opt/elasticsearch/bin/elasticsearch"
pid_file="/var/run/$name.pid"
NICE_LEVEL="-n 19"
start () {
    command="/usr/bin/nice ${NICE_LEVEL} ${elastic}"

    log_daemon_msg "Starting $mode" "$name"
    if start-stop-daemon --start --chuid "beholder" --quiet --oknodo --pidfile "$pid_file" -b -m --exec $command; then
        log_end_msg 0
    else
        log_end_msg 1
    fi
}
stop () {
    echo "Stopping $name"
    start-stop-daemon --stop --quiet --oknodo --pidfile "$pid_file"
    echo "$name stopped"
}

status () {
    status_of_proc -p $pid_file "" "$name"
}
case $1 in
    start)
        if status; then exit 0; fi
        start
        ;;
    stop)
        stop
        ;;
    reload)
        stop
        start
        ;;
    restart)
        stop
        start
        ;;
    status)
        status && exit 0 || exit $?
        ;;
    *)
        echo "Usage: $0 {start|stop|restart|reload|status}"
        exit 1
        ;;
esac
exit 0
EOF
chmod +x elasticsearch
update-rc.d elasticsearch defaults
############
#Kibana Init
############
cd /etc/init.d
cat <<'EOF' > kibana
. /lib/lsb/init-functions
name="kibana"
kibana="-- /opt/kibana/bin/kibana"
pid_file="/var/run/$name.pid"
NICE_LEVEL="-n 19"
start () {
    command="/usr/bin/nice ${NICE_LEVEL} ${kibana}"

    log_daemon_msg "Starting $mode" "$name"
    if start-stop-daemon --start --chuid "beholder" --quiet --oknodo --pidfile "$pid_file" -b -m --exec $command; then
        log_end_msg 0
    else
        log_end_msg 1
    fi
}
stop () {
    echo "Stopping $name"
    start-stop-daemon --stop --quiet --oknodo --pidfile "$pid_file"
    echo "$name stopped"
}

status () {
    status_of_proc -p $pid_file "" "$name"
}
case $1 in
    start)
        if status; then exit 0; fi
        start
        ;;
    stop)
        stop
        ;;
    reload)
        stop
        start
        ;;
    restart)
        stop
        start
        ;;
    status)
        status && exit 0 || exit $?
        ;;
    *)
        echo "Usage: $0 {start|stop|restart|reload|status}"
        exit 1
        ;;
esac
exit 0
EOF
chmod +x kibana
update-rc.d kibana defaults
#################
#Fixing the Crons
#################
echo "[+] Fixin your crons!"
cd /opt/
cat <<EOF > cron
0-59/5 * * * * /opt/bro/bin/broctl cron
0 0 * * * /usr/local/bin/curator --host localhost --port 9200 close indices --older-than 30 --time-unit days --timestring '%Y.%m.%d' --prefix bro
EOF
crontab cron
rm -rf cron
######################
#CHOWNing your system!
######################
chown -R beholder:beholder /logs
chown -R beholder:beholder /opt
######################
#Clearing Certificates
######################
update-ca-certificates -f
###################
#Apache - a2e Setup
###################
echo "[+] Setting up Apache."
a2enmod proxy proxy_http ssl
##################################
#Apache - Creating Basic Auth User
##################################
htpasswd -cbm /etc/apache2/.htpasswd beholder beholder
###################################
#Apache - Creating self-signed cert
###################################
openssl req -x509 -nodes -days 365 -newkey rsa:2048 -subj "/C=DD/ST=Guarding/L=Caverns/O=beholder/CN=beholder" -keyout /etc/ssl/certs/beholder.key -out /etc/ssl/certs/beholder.crt
################################
#Apache - A2E Enable and Disable
################################
a2ensite default-ssl
a2dissite 000-default
########################
#Apache - Port Listening
########################
cat <<EOF > /etc/apache2/ports.conf
Listen 443
EOF
#########################
#Apache - Sites-Available
#########################
cat <<EOF > /etc/apache2/sites-available/default-ssl.conf
<IfModule mod_ssl.c>
        <VirtualHost *:443>
                ServerAdmin webmaster@localhost
                <Proxy *>
                Order deny,allow
                Allow from all
                AuthType Basic
                AuthName "Access Kibana"
                AuthUserFile /etc/apache2/.htpasswd
                Require valid-user
                </Proxy>
                ProxyPass / http://localhost:5601/
                ProxyPassReverse / http://localhost:5601/
                SSLEngine on
                SSLCertificateFile      /etc/ssl/certs/beholder.crt
                SSLCertificateKeyFile   /etc/ssl/certs/beholder.key
                <FilesMatch "\.(cgi|shtml|phtml|php)$">
                                SSLOptions +StdEnvVars
                </FilesMatch>
                <Directory /usr/lib/cgi-bin>
                                SSLOptions +StdEnvVars
                </Directory>
                BrowserMatch "MSIE [2-6]" \
                                nokeepalive ssl-unclean-shutdown \
                                downgrade-1.0 force-response-1.0
                # MSIE 7 and newer should be able to use keepalive
                BrowserMatch "MSIE [17-9]" ssl-unclean-shutdown
        </VirtualHost>
</IfModule>
EOF
#############
#Firewall Fix
#############
echo "[+] Configuring the firewall."
ufw deny 5601
ufw default allow
ufw enable
#########
#Finished
#########
clear
echo 'Your installation has finished. We are rebooting your system.'
seconds=10
while [ $seconds -gt 0 ];
        do
                echo "$seconds"
                sleep 1s
                seconds=$(($seconds - 1))
        done
echo 'VGhhbmsgeW91IGZvciB0cnlpbmcgdGhlIEJlaG9sZGVyIHNjcmlwdCEgVGhlIHBhbmNha2VzIGFyZSBub3QgYSBsaWUu'
shutdown -r now